docs/proposal-new-output-syntax.md

```diff
@@ -0,0 +1,765 @@
+```
+
+Note that an artificial dependency can be used to *ensure that at least one
+of the alternate paths gets taken*:
+```
+a:x? => b1
+a:y? => b2
+b1 | b2 => c  # join
+a => c # ARTIFICIAL DEPENDENCY
+```
+This forces `c` to be spawned as an unsatisfied prerequisite waiting on `b1` or
+`b2` even if neither branch runs. If the alternate paths don't actually need to
+join, `c` can be a dummy task.
+
+*See Appendix Output Completion Expressions for a possible follow-up that
+would allow us to flag `a` as incomplete here and avoid the artificial
+dependency.*
+
+## Failed Tasks
+
+No special handling of failed tasks is required.
+- if success is required, a failed task automatically gets retained in the task
+  pool as incomplete.
+- if success is optional (which would usually imply that optional failure is
+  explicitly handled), a failed task can be removed as complete
+
+No special syntax (e.g. `c:fail => NULL`) or suicide triggers (e.g. `c:fail =>
+!c`) are needed for leaf tasks either, if we don't care if they succeed or fail:
+```
+a => b => c?  # success of c is optional
+```
+
+## Family Triggers
+
+Conceivably we could try to interpret the output signifier in terms of a
+collective pseudo-output for the family as whole, but that gets really hard to
+understand.
+
+Fortunately, the output type signifier can simply be applied directly to all
+family members, for all family trigger types. That's easy to understand,
+and it makes sense because the signifier doesn't affect triggering, it only
+affects what we do with family members if the outputs in question are not
+completed (i.e. keep them in n=0 as incomplete, or forget them).
+
+If particular member outputs need special treatment they should be singled
+out in the graph - just like special member triggers.
+
+```
+A:succeed-all => b
+```
+- trigger b only if all members succeed; and all members are required to
+  succeed
+
+```
+A:succeed-all? => b
+```
+- trigger b only if all members succeed; but it's OK if they fail (maybe the
+  failed path is handled elsewhere)
+
+```
+A:succeed-any => b
+```
+- trigger b if any member succeeds; but any member that runs is required to
+  succeed
+
+```
+A:succeed-any? => b
+```
+- trigger b if any member succeeds; but it's OK if any or all members fail
+
+
+## Backward Compatibility (Cylc 7)
+
+In Cylc 7 graphs, all outputs are effectively "required" in that all tasks are
+pre-spawned to the next cycle point, and users are expected to define suicide
+triggers to remove unused branches that are really optional.
+
+In Cylc 8, suicide triggers mostly aren't needed because unused branches don't
+get spawned at all, but they're still there and still if needed.
+
+So if a Cylc 7 workflow is detected (via `suite.rc` vs `flow.cylc` config
+filename):
+- instead of flagging use of both `foo:succeed` and `foo:fail` as an error,
+we can infer that success must be optional for `foo`, to avoid ending up
+with an incomplete task due to the unused output
```

Comment:
Currently users get a warning that the `suite.rc` filename is deprecated, so if they rename their `suite.rc` to `flow.cylc`, they will lose this compatibility. Perhaps we will need a Cylc 7 compatibility mode based on `suite.rc` detection that loudly warns about the graphing changes, and that the changes will need to be made if renaming to `flow.cylc`. Or will it be better to just infer optional success for `flow.cylc` workflows as well, with a warning that this inference is deprecated?

Reply:
We're going to have to distinguish between Cylc 7 and 8 workflows for cylc/cylc-flow#2324 as well so we need to treat `suite.rc` and `flow.cylc` files differently. I agree we'll need to review the deprecation message.

Reply:
Noted this as unresolved in the proposal, with another possible option.